jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,210 @@
+            return Optional.empty();
+          }
+        } catch (DateTimeParseException | IOException ex) {
+          // If reading update time failed, file might be corrupt, so delete it
+          Files.delete(lastUpdateCheck);
+          log.accept(LogEvent.debug("Failed to read lastUpdateCheck; " + ex.getMessage()));
+        }
+      }
+
+      // Check for update
+      HttpURLConnection connection = (HttpURLConnection) new URL(versionUrl).openConnection();
+      try {
+        connection.setConnectTimeout(3000);
+        BufferedReader bufferedReader =
+            new BufferedReader(
+                new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
+        String latestVersion = bufferedReader.readLine().trim();
+        Files.write(lastUpdateCheck, Instant.now().toString().getBytes(StandardCharsets.UTF_8));
+        if (currentVersion.equals(latestVersion)) {
+          return Optional.empty();
+        }
+        return Optional.of(
+            "A new version of Jib ("
+                + latestVersion
+                + ") is available (currently using "
+                + currentVersion
+                + "). Update your build configuration to use the latest features and fixes!");
+
+      } finally {
+        connection.disconnect();
+      }
+
+    } catch (IOException ex) {
+      log.accept(LogEvent.debug("Update check failed; " + ex.getMessage()));
+    }
+
+    return Optional.empty();
+  }
+
+  /**
+   * Returns the config directory set by {@link PropertyNames#CONFIG_DIRECTORY} if not null,
+   * otherwise returns the default config directory.
+   *
+   * @return the config directory set by {@link PropertyNames#CONFIG_DIRECTORY} if not null,
+   *     otherwise returns the default config directory.
+   */
+  private static Path getConfigDir() {
+    String configDirProperty = System.getProperty(PropertyNames.CONFIG_DIRECTORY);
+    if (!Strings.isNullOrEmpty(configDirProperty)) {
+      return Paths.get(configDirProperty);
+    }
+    return XdgDirectories.getConfigHome();
+  }
+
+  private final Future<Optional<String>> updateMessageFuture;
+
+  @VisibleForTesting
+  UpdateChecker(Future<Optional<String>> updateMessageFuture) {
+    this.updateMessageFuture = updateMessageFuture;
+  }
+
+  /**
+   * Returns a message indicating Jib should be upgraded if the check succeeded and the current
+   * version is outdated, or returns {@code Optional.empty()} if the check was interrupted or did
+   * not determine that a later version was available.
+   *
+   * @return the {@link Optional} message to upgrade Jib if a later version was found, else {@code
+   *     Optional.empty()}.
+   */
+  public Optional<String> finishUpdateCheck() {
+    if (updateMessageFuture.isDone()) {
+      try {
+        return updateMessageFuture.get();
+      } catch (InterruptedException | ExecutionException ignored) {
+        // Fail silently;
+      }
+    }
+    updateMessageFuture.cancel(true);
```

Comment:
can we close the connection here to interrupt a read?

Reply:
Is it necessary? `updateMessageFuture.cancel(true)` should interrupt the read regardless, shouldn't it? Or will the connection be left open if we interrupt it and don't disconnect here?

Reply:
I'm not really concerned about this. Looks like the Maven/Gradle JVM always terminates no matter what I do in the update checker thread. And particularly it doesn't matter much to close the connection if the build is about to exit.

Reply:
I think the 3 second timeout is generally enough.